File: Structy/Hashing/Anagrams.java
import java.util.HashMap;

  
class Source {

  public static HashMap<Character, Integer> charCount(String s){
    HashMap<Character, Integer> count = new HashMap<Character, Integer>();

    for(char c : s.toCharArray()){

      if(count.get(c) == null){
        count.put(c, 0);
      }
      count.put(c, count.get(c) + 1);
    }
    return count; 
  }
  
  public static boolean anagrams(String s1, String s2) {
    
    return charCount(s1).equals(charCount(s2));
  }

  public static void run() {
    // this function behaves as `main()` for the 'run' command
    // you may sandbox in this function , but should not remove it
  }
}

/*
    n = length of s1
    m = length of s2
    time complexity O(n + m)
    space complexity O(n+m)
*/

File: Structy/Python/Hashing/Anagrams.py
def anagrams(s1, s2):
  return char_count(s1) == char_count(s2)

def char_count(s):
  count = {}
  for char in s:
    if char not in count:
      count[char] = 0
    count[char] += 1
    
  return count

# n is length of s1 and m is length of s2
#Time Complexity: O(n + m)
#Space Complexity: O(n + m) 
#Reason: We are using a dictionary to store the character counts.
#Reason: We are iterating through the strings to count the characters.  